orc/src/main/java/org/apache/iceberg/orc/OrcFileAppender.java

```diff
@@ -99,9 +104,27 @@ public Metrics metrics() {
 
   @Override
   public long length() {
-    Preconditions.checkState(isClosed,
-        "Cannot return length while appending to an open file.");
-    return file.toInputFile().getLength();
+    if (isClosed) {
+      return file.toInputFile().getLength();
+    }
+    if (this.treeWriter == null) {
+      throw new RuntimeException("Can't get the length!");
+    }
+    long estimateMemory = this.treeWriter.estimateMemory();
+
+    long dataLength = 0;
+    try {
+      List<StripeInformation> stripes = writer.getStripes();
+      if (!stripes.isEmpty()) {
+        StripeInformation stripeInformation = stripes.get(stripes.size() - 1);
+        dataLength = stripeInformation != null ? stripeInformation.getOffset() + stripeInformation.getLength() : 0;
+      }
+    } catch (IOException e) {
+      throw new UncheckedIOException(String.format("Can't get stripes from file %s", file.location()), e);
+    }
+
+    // This value is estimated, not actual.
+    return dataLength + estimateMemory + batch.size;
```

Comment:
I read this https://github.com/apache/iceberg/pull/3784#issuecomment-1022891787 here. For the file-persisted bytes , I think using the last strip's offset plus the its length should be correct.  For he the memory encoded batch vector , I think the `TreeWriter#estimateMemory` should be okay. 

But for the batch vector whose rows did not flush to encoded memory,  using the batch.size shouldn't be correct. Because the rows can be any data type,  such as Integer, Long, Timestamp, String etc. As their width are  not the same, I think we may need to use an average width minus the batch.size (which is row count actually).